src/Val.cc

```diff
@@ -2717,19 +2715,25 @@ unsigned int TableVal::ComputeFootprint(std::unordered_set<const Val*>* analyzed
 }
 
 std::unique_ptr<detail::HashKey> TableVal::MakeHashKey(const Val& index) const {
-    return table_hash->MakeHashKey(index, true);
+    return GetTableHash()->MakeHashKey(index, true);
 }
 
-void TableVal::SaveParseTimeTableState(RecordType* rt) {
+std::set<TableType*> TableVal::SaveParseTimeTableState(RecordType* rt) {
+    std::set<TableType*> affected_table_types;
+
     auto it = parse_time_table_record_dependencies.find(rt);
 
     if ( it == parse_time_table_record_dependencies.end() )
-        return;
+        return affected_table_types;
 
     auto& table_vals = it->second;
 
-    for ( auto& tv : table_vals )
+    for ( auto& tv : table_vals ) {
         parse_time_table_states[tv.get()] = tv->DumpTableState();
+        affected_table_types.insert(tv->table_type.get());
```

Comment:
Matter of taste I guess, but I'd have tended to leave `SaveParseTimeTableSate()` and `RecordType::AddFields()` unmodified and add an extra loop in `RebuildParseTimeTables()`:
```
 
 void TableVal::RebuildParseTimeTables() {
+    std::set<TableType*> affected_table_types;
+    for ( auto& [tv, _] : parse_time_table_states )
+        affected_table_types.insert(tv->table_type.get());
+
+    for ( auto* tt : affected_table_types )
+        tt->RegenerateHash();
+
     for ( auto& [tv, ptts] : parse_time_table_states )
         tv->RebuildTable(std::move(ptts));
```

Don't think runtime matters at this stage and easier to justify getting away without comments what is returned and why there's the hash regeneration in the middle of `AddFields()` now ;-)

Reply:
> Matter of taste I guess, but I'd have tended to leave `SaveParseTimeTableSate()` and `RecordType::AddFields()` unmodified and add an extra loop in `RebuildParseTimeTables()`

I'm completely happy for you to restructure that too if you wish!